### PR TITLE
#170947670 Reject a trip request 

### PR DESCRIPTION
**Pull request by T2Wil**
#### What does this PR do?
This PR will enable a manager to reject the request made.
#### Description of Task to be completed?
- authenticate manager
- find the request
- change the request' status to `rejected`
- write tests
- write documentation

#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `cd knights-bn-backend` to go to the project directory.
- run `git checkout ft-reject-request-170947670` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres database locally.
- run `npm run createTables` to create the database tables and seed in primary data.
- create a .env file, paste in below variables and replace xxxx with the correct values: (Check code-knights `slack channel` to find the correct values)

```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
GOOGLE_CLIENT_ID = 'XXXXX'
GOOGLE_CLIENT_SECRET = 'XXXXX'
GOOGLE_CALLBACK_URL = 'XXXXX'
FCBK_CLIENT_ID = 'XXXXX'
FCBK_CLIENT_SECRET = 'XXXXX'
FCBK_CALLBACK_URL = 'XXXXX'
BN_API_KEY= 'XXXX'
MANAGER_PSWD= 'XXXX'
```
- run `npm dev-start` to start the server.
- open postman and do sign up to acquire the token `PATCH` `http://localhost:4000/api/v1/trips/reject`
- in the query params create a key-value pair: where `requestId` is a key and its value has to be the one of requestId you want to reject as a `manager`
as : 
image

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[link](https://www.pivotaltracker.com/story/show/170947670)

##### 1. do POST `localhost:4000/api/v1/auth/signup`  to register a new user (a requester by default)
new user info: 
```
{
"firstName": "John",
"lastName": "Doe",
"gender": "male",
"passportNumber": "123482111",
"email": "johndoe@gmail.com",
"password": "Password@1"
}

```
image

##### 2. do POST `localhost:4000/api/v1/trips/returnTrip` As a user (John Doe) to request a two-way trip.
Trip info:
```
{
	"origin": "Kigali",
	"destination": "Kampala",
	"departureDate": "2020-03-04",
	"returnDate": "2020-04-29",
	"accommodation": "XYZ campus",
	"reason": "professional meetup"
}
```
image

John's request is recorded under an ID of 4. Make sure to keep track of this id as it might differ on your end.

##### 3. do POST `localhost:4000/api/v1/auth/login` and log in as a manager `eugene`; 
Manager credentials
```
{
        "email":"eugene.munyampundu@gmail.com",
	"password":"Niyonkuru@1"
}
```
In the response body: you should see that `Eugene` 's role is manager.

[image](https://user-images.githubusercontent.com/30593186/75610970-cb872680-5b1e-11ea-8838-afb9096eb4d4.png)

##### 5. do GET 'localhost:4000/api/v1/trips/pendingApproval` to view all available requests under manager `eugene`. 
image

Make sure to find the request-id of the requester`Eugene` on my end is `4`.

##### 4. do PATCH 'localhost:4000/api/v1/trips/reject`  As a manager `Eugene` give the requestId query`John`'s requestId as illustrated below:
image

##### 5. As a manager, when I try to reject a rejected trip request:
image

### CODE REVIEW

**Comment by Kagorora** [comment]
@T2Wil  How as a manager, I should know the request ID?, so that I can be able to reject them, I suggest That you consider an option to first see the available request assigned to me(as manager) then make a decision later. 

As an example, I can't tell if requestId 2 exists
Screen Shot 2020-02-29 at 17 09 17

**Comment by T2Wil** [comment]
> @T2Wil How as a manager, I should know the request ID?, so that I can be able to reject them, I suggest That you consider an option to first see the available request assigned to me(as manager) then make a decision later.
> 
> As an example, I can't tell if requestId 2 exists
> Screen Shot 2020-02-29 at 17 09 17

Thanks for the feedback @Kagorora , I missed the step but now It is included.
Before rejecting a request, a manager might first check out the available requests.

**Comment by Kagorora** [comment]
> > @T2Wil How as a manager, I should know the request ID?, so that I can be able to reject them, I suggest That you consider an option to first see the available request assigned to me(as manager) then make a decision later.
> > As an example, I can't tell if requestId 2 exists
> > Screen Shot 2020-02-29 at 17 09 17
> 
> Thanks for the feedback @Kagorora , I missed the step but now It is included.
> Before rejecting a request, a manager might first check out the available requests.

>> @T2Wil  Thank you for your consideration, now I can see all my(as a manager) assigned request then reject accordingly, nice work

**Review by Kagorora — APPROVED** [review]

**Review by Niyonsengaeric — APPROVED** [review]

File: src/controllers/request.js [hunk 1]
```diff
@@ -95,4 +97,99 @@ export default class usersController {
...
+        })
+        if(managerExists){
+            const request = await models.Request.findAll(
+                {
+                    where: {
+                        id: requestId,
+                        managerId
+                    }
+                }
+            ).then ((request) => {
```
Comment (kagabof) [inline src/controllers/request.js]: There is no reason for using await and then at the same time, please change that.

File: src/controllers/request.js [hunk 2]
```diff
@@ -95,4 +97,99 @@ export default class usersController {
...
+            
+            
+            if(request.length !== 0){
+                let requestStatus = request[0].status;
+
+                if(requestStatus === 'pending'){
+                    const updatedRequest = await models.Request.update(
+                        { status: 'rejected' },
+                        { where : { id: requestId } }
+                    ).then((request)=> {
```
Comment (kagabof) [inline src/controllers/request.js]: There is no reason for using await and then at the same time, please change that.

File: src/controllers/request.js [hunk 3]
```diff
@@ -95,4 +97,99 @@ export default class usersController {
...
+                    if(tripStarted){
+                        res.status(405).json({
+                            message: "Sorry can't reject ! The user is now on trip."
+
+                        })
+                    } else {
+                        const updatedRequest = await models.Request.update(
+                            { status: 'rejected' },
+                            { where : { id: requestId } }
+                        ).then((request)=> {
```
Comment (kagabof) [inline src/controllers/request.js]: There is no reason for using await and then at the same time, please change that.

File: src/tests/rejectRequest.test.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+const testRejectRequest = () => {
+  describe('Test /trips/reject', () => {
+    it('returns 200 on a user signed in successfully', (done) => {
+      chai
+        .request(app)
+        .post('/api/v1/auth/login')
+        .send(mockData.loginX)
+        .end((err, res) => {
+          expect(res.statusCode).to.equal(200);
```
Comment (kagabof) [inline src/tests/rejectRequest.test.js:21]: please test at list the message.

File: src/tests/rejectRequest.test.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    it('returns 200 on a manager signed in successfully', (done) => {
+      chai
+        .request(app)
+        .post('/api/v1/auth/login')
+        .send({
+          email: 'eugene.munyampundu@gmail.com',
+          password: 'Niyonkuru@1'
+        })
+        .end((err, res) => {
+          expect(res.statusCode).to.equal(200);
```
Comment (kagabof) [inline src/tests/rejectRequest.test.js:45]: please test at list the message.

**Review by kagabof — CHANGES_REQUESTED** [review]

File: src/controllers/request.js (hunk 1 above)
Comment (T2Wil) [reply src/controllers/request.js]: Thanks @kagabof , 
I have learnt it.

File: src/controllers/request.js (hunk 2 above)
Comment (T2Wil) [reply src/controllers/request.js]: corrected!

File: src/controllers/request.js (hunk 3 above)
Comment (T2Wil) [reply src/controllers/request.js]: Corrected.

File: src/tests/rejectRequest.test.js (hunk 4 above)
Comment (T2Wil) [reply src/tests/rejectRequest.test.js:21]: Added some!
Thanks @kagabof

File: src/tests/rejectRequest.test.js (hunk 5 above)
Comment (T2Wil) [reply src/tests/rejectRequest.test.js:45]: Added tests.

**Pull request merged**